### ISSUE TITLE
Cannot use ENV var processors in the adapter configuration

### ISSUE THREAD

**Issue by yakobe** [1up-lab/OneupFlysystemBundle#281]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | yes
| Version     | 4.8.0

#### Summary

Before the change in https://github.com/1up-lab/OneupFlysystemBundle/pull/277 it was possible to  resolve environment variables in the configuration. E.g. 

```yaml
oneup_flysystem:
  adapters:
    local_adapter:
      local:
        location: '%env(resolve:APP_DATA_DIR)%'
```
And the `APP_DATA_DIR` could then be  set to`%kernel.project_dir%/var/data`

However, now is is not possible and yields the error:
```
!!  In MergeExtensionConfigurationPass.php line 200:
!!
!!    Using a cast in "env(resolve:APP_DATA_DIR)" is incompatible with resolution
!!     at compile time in "Oneup\FlysystemBundle\DependencyInjection\OneupFlysyst
!!    emExtension". The logic in the extension should be moved to a compiler pass
!!    , or an env parameter with no cast should be used instead.
```

I guess it is not possble to use the [environment variable processors](https://symfony.com/doc/current/configuration/env_var_processors.html) now because the configuration needs to be compiled first so that the adapter itself can be changed with an environment variable.

#### How to reproduce

Configure a local adapter and set it's location to `location: '%env(resolve:APP_DATA_DIR)%'`. Then add `%kernel.project_dir%/var/data` to your `.env.local` file

**Comment by bytehead** [comment]
Thanks for reporting! I get your point, need to think about it. As #277 is the only new thing in 4.8.0, I recommend to use 4.7.0 until we have a solution here.

**Comment by alexandre-le-borgne** [comment]
(translated)

The pull request I made works in my case (the adapter used depends on the environment). I had taken the line I added from another vendor (I can't remember which one, probably an official symfony component) without thinking too much about it.

resolveEnvPlaceholders is probably not a real fix. I haven't taken the time to find a better fix and don't need one at the moment. Maybe someone who knows more about DependencyInjection configuration could help.

**Comment by alexandre-le-borgne** [comment]
(translated) 

After a bit of reading about the obscure workings of Symfony's dependency injection component, I'm beginning to think that my PR may not be the right solution according to Nicolas Grekas & co. However, they don't seem to offer an official solution for this case...

I could have used a parameters.local.yaml, ignored by git and imported into services.yaml; allowing me to override the configuration without the environment variables. 
A solution that would also concern this issue, as the "location" is not something you want to change at runtime.

**Comment by yakobe** [comment]
We do something a bit similar to that to switch adapters in prod/test/dev etc.

We have a parameter for the service name:
```yaml
parameters:
  filesystem_service: 'oneup_flysystem.app_data_%env(DATA_STORE)%_filesystem'
```

and then we pass that service to a wrapper that manages our file system interactions

```yaml
services:
  App\FileSystem\FilesystemOperator:
    arguments:
      $filesystem: '@=service(parameter(''filesystem_service''))'
```

Not perfect, but it does the job

**Comment by bytehead** [comment]
I'm afraid we'll have to revert the changes from #277 for the moment now.

**Comment by MarieMinasyan** [comment]
Hi,

Since the latest changes a simple configuration like this doesn't work anymore because the environment variables are resolved on build and not on runtime as they're supposed to be.
```
oneup_flysystem:
    adapters:
        attachments:
            awss3v3:
                client: app.storage.aws_s3_client
                bucket: "%env(AWS_S3_BUCKET)%"
                prefix: "tes/attachments"
```

**Comment by bytehead** [comment]
See https://github.com/1up-lab/OneupFlysystemBundle/issues/281#issuecomment-1589217876

**Comment by kosmala007** [comment]
After upgrading to 4.8 I am not able to use the FTP adapter, because I was getting an error

```
Using a cast in "env(int:CDN_FTP_PORT)" is incompatible with resolution at compile time in "Oneup\FlysystemBundle\DependencyInjection\OneupFlysystemExtension". The logic in the extension should be moved to a compiler pass, or an env parameter with no cast should be used instead.
```

this is my config:

```yaml
oneup_flysystem:
   # ...
   port: '%env(int:CDN_FTP_PORT)%'
```

```env
CDN_FTP_PORT=21
```

and I can't use the string for CDN_FTP_PORT because ``League\Flysystem\FtpConnectionOptions::__construct'' only accepts an int for port

I had to downgrade the package to be able to use FTP adapter

**Comment by bytehead** [comment]
I've reverted the failing changes in #285 and released `4.9.0`.
